src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -979,7 +1001,22 @@ public GHPermissionType fetch(String username)
                 }
               });
 
-          if (request.isFetchBranches() && !request.isComplete()) {
+          // JENKINS-65071
+          // Observer has information about the events to analyze. To avoid unnecessary processing
+          // and GitHub API requests,
+          // it is necessary to check if this event contains a set of {@link SCMHead} instances of a
+          // type.
+          // When we open or close a Pull request we don't need a TAG examination because the event
+          // doesn't have any TAG. So, we only trigger a
+          // examination if the observer has any include event of each type BranchSCMHead,
+          // PullRequestSCMHead or GitHubTagSCMHead.
+          // But when a project scan is triggered we don't have any event so a full examination
+          // should happen.
+
+          if (request.isFetchBranches()
+              && !request.isComplete()
+              && (event == null
+                  || this.checkObserverIncludesType(observer.getIncludes(), BranchSCMHead.class))) {
```

Comment:
I feel like there should be a method for this.  Either instead of or inaddition to `checkObserverIncludesType`, please create a method `shouldRetrieve(@NonNull SCMHeadObserver observer,
      @CheckForNull SCMHeadEvent<?> event, @NonNull Class t)`.  I'm open to a different name, but I want the logic described in this comment encapsulated in that method.   Thjen this comment goes in there and it can be called three times from this method. 

Then your tests can cover more of the logic without calling `retrieve()`. This will also ensure this important description of the behavior doesn't get lost.